utils.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+)
+
+var (
+	kubeletChainsRegex = regexp.MustCompile(`(?m)^:(KUBE-IPTABLES-HINT|KUBE-KUBELET-CANARY)`)
+	ruleEntryRegex     = regexp.MustCompile(`(?m)^-`)
+)
+
+// hasKubeletChains checks if the output of an iptables*-save command
+// contains any of the rules set by kubelet.
+func hasKubeletChains(output []byte) bool {
+	return kubeletChainsRegex.Match(output)
+}
+
+// ruleEntriesNum counts how many rules there are in an iptables*-save command
+// output.
+func ruleEntriesNum(iptablesOutput []byte) int {
+	return len(ruleEntryRegex.FindAllIndex(iptablesOutput, -1))
+}
+
+// logErrorAndExit prints an error message to stderr and exits the program with
+// an error code.
+func logErrorAndExit(err error) {
+	fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+	os.Exit(1)
+}
+
+// detectIptablesDir tries to detect the `iptables` location in
+// either /usr/sbin or /sbin. If it's not there, it returns an error.
+func detectIptablesDir() (string, error) {
+	if fileExists("/usr/sbin/iptables") {
+		return "/usr/sbin", nil
+	} else if fileExists("/sbin/iptables") {
+		return "/sbin", nil
+	} else {
+		return "", errors.New("iptables is not present in either /usr/sbin or /sbin")
+	}
+}
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return err == nil
+}
+
+// executableExists checks if a file exists and it's executable by someone.
+func executableExists(path string) bool {
+	stat, err := os.Stat(path)
+	return err == nil && stat.Mode()&0o111 != 0
+}
+
+// runAndReadError runs a exec.Cmd and tries to extract the error message from stderr
+// if present and it includes it in the returned error. This overrides the Stderr in cmd if
+// present.
+func runAndReadError(cmd *exec.Cmd) error {
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+
+	if err := cmd.Run(); err != nil {
+		if stderr.Len() > 0 {
+			err = fmt.Errorf("%s: %v", stderr.String(), err)
```

Comment:
I think in general you want _either_ `err` or `stderr.String()`, but not both

Reply:
err can have info not contained by stderr, like the exit code for example (for ExitError errors)